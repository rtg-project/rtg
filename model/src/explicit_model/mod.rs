pub mod entity;
pub mod field;
pub mod json_schema;
pub mod model;
pub mod sql_type;

// Tests
#[cfg(test)]
mod tests {

  use similar_asserts::assert_eq;
  #[test]
  fn it_works() {
    assert_eq!(2 + 2, 4);
  }

  #[test]
  fn it_works2() {
    assert_eq!(3 + 2, 5);
  }
}
