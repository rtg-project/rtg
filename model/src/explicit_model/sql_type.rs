// See rust-postgres
use derive_more::Display;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

// See https://github.com/sfackler/rust-postgres/blob/fc10985f9fdf0903893109bc951fb5891539bf97/postgres-types/src/type_gen.rs
#[derive(Serialize, Deserialize, Debug, PartialEq, JsonSchema, Clone, Display)]
#[serde(rename_all = "camelCase")]
pub enum Type {
  #[display(fmt = "bool")]
  Bool,
  #[display(fmt = "bytea")]
  Bytea,
  #[display(fmt = "char")]
  Char,
  #[display(fmt = "name")]
  Name,
  #[display(fmt = "int8")]
  Int8,
  #[display(fmt = "int2")]
  Int2,
  #[display(fmt = "int2vector")]
  Int2Vector,
  #[display(fmt = "int4")]
  Int4,
  #[display(fmt = "regproc")]
  Regproc,
  #[display(fmt = "text")]
  Text,
  #[display(fmt = "oid")]
  Oid,
  #[display(fmt = "tid")]
  Tid,
  #[display(fmt = "xid")]
  Xid,
  #[display(fmt = "cid")]
  Cid,
  #[display(fmt = "oidvector")]
  OidVector,
  #[display(fmt = "pgddlcommand")]
  PgDdlCommand,
  #[display(fmt = "json")]
  Json,
  #[display(fmt = "xml")]
  Xml,
  #[display(fmt = "xml[]")]
  XmlArray,
  #[display(fmt = "pgnodetree")]
  PgNodeTree,
  #[display(fmt = "json[]")]
  JsonArray,
  #[display(fmt = "tableamhandler")]
  TableAmHandler,
  #[display(fmt = "xid8[]")]
  Xid8Array,
  #[display(fmt = "indexamhandler")]
  IndexAmHandler,
  #[display(fmt = "point")]
  Point,
  #[display(fmt = "lseg")]
  Lseg,
  #[display(fmt = "path")]
  Path,
  #[display(fmt = "box")]
  Box,
  #[display(fmt = "polygon")]
  Polygon,
  #[display(fmt = "line")]
  Line,
  #[display(fmt = "line[]")]
  LineArray,
  #[display(fmt = "cidr")]
  Cidr,
  #[display(fmt = "cidr[]")]
  CidrArray,
  #[display(fmt = "float4")]
  Float4,
  #[display(fmt = "float8")]
  Float8,
  #[display(fmt = "unknown")]
  Unknown,
  #[display(fmt = "circle")]
  Circle,
  #[display(fmt = "circle[]")]
  CircleArray,
  #[display(fmt = "macaddr8")]
  Macaddr8,
  #[display(fmt = "macaddr8[]")]
  Macaddr8Array,
  #[display(fmt = "money")]
  Money,
  #[display(fmt = "money[]")]
  MoneyArray,
  #[display(fmt = "macaddr")]
  Macaddr,
  #[display(fmt = "inet")]
  Inet,
  #[display(fmt = "bool[]")]
  BoolArray,
  #[display(fmt = "bytea[]")]
  ByteaArray,
  #[display(fmt = "char[]")]
  CharArray,
  #[display(fmt = "name[]")]
  NameArray,
  #[display(fmt = "int2[]")]
  Int2Array,
  #[display(fmt = "int2vector[]")]
  Int2VectorArray,
  #[display(fmt = "int4[]")]
  Int4Array,
  #[display(fmt = "regproc[]")]
  RegprocArray,
  #[display(fmt = "text[]")]
  TextArray,
  #[display(fmt = "tid[]")]
  TidArray,
  #[display(fmt = "xid[]")]
  XidArray,
  #[display(fmt = "cid[]")]
  CidArray,
  #[display(fmt = "oidvector[]")]
  OidVectorArray,
  #[display(fmt = "bpchar[]")]
  BpcharArray,
  #[display(fmt = "varchar[]")]
  VarcharArray,
  #[display(fmt = "int8[]")]
  Int8Array,
  #[display(fmt = "point[]")]
  PointArray,
  #[display(fmt = "lseg[]")]
  LsegArray,
  #[display(fmt = "path[]")]
  PathArray,
  #[display(fmt = "box[]")]
  BoxArray,
  #[display(fmt = "float4[]")]
  Float4Array,
  #[display(fmt = "float8[]")]
  Float8Array,
  #[display(fmt = "polygon[]")]
  PolygonArray,
  #[display(fmt = "oid[]")]
  OidArray,
  #[display(fmt = "aclitem")]
  Aclitem,
  #[display(fmt = "aclitem[]")]
  AclitemArray,
  #[display(fmt = "macaddr[]")]
  MacaddrArray,
  #[display(fmt = "inet[]")]
  InetArray,
  #[display(fmt = "bpchar")]
  Bpchar,
  #[display(fmt = "varchar")]
  Varchar,
  #[display(fmt = "date")]
  Date,
  #[display(fmt = "time")]
  Time,
  #[display(fmt = "timestamp")]
  Timestamp,
  #[display(fmt = "timestamp[]")]
  TimestampArray,
  #[display(fmt = "date[]")]
  DateArray,
  #[display(fmt = "time[]")]
  TimeArray,
  #[display(fmt = "timestamptz")]
  Timestamptz,
  #[display(fmt = "timestamptz[]")]
  TimestamptzArray,
  #[display(fmt = "interval")]
  Interval,
  #[display(fmt = "interval[]")]
  IntervalArray,
  #[display(fmt = "numeric[]")]
  NumericArray,
  #[display(fmt = "cstring[]")]
  CstringArray,
  #[display(fmt = "timetz")]
  Timetz,
  #[display(fmt = "timetz[]")]
  TimetzArray,
  #[display(fmt = "bit")]
  Bit,
  #[display(fmt = "bit[]")]
  BitArray,
  #[display(fmt = "varbit")]
  Varbit,
  #[display(fmt = "varbit[]")]
  VarbitArray,
  #[display(fmt = "numeric")]
  Numeric,
  #[display(fmt = "refcursor")]
  Refcursor,
  #[display(fmt = "refcursor[]")]
  RefcursorArray,
  #[display(fmt = "regprocedure")]
  Regprocedure,
  #[display(fmt = "regoper")]
  Regoper,
  #[display(fmt = "regoperator")]
  Regoperator,
  #[display(fmt = "regclass")]
  Regclass,
  #[display(fmt = "regtype")]
  Regtype,
  #[display(fmt = "regprocedure[]")]
  RegprocedureArray,
  #[display(fmt = "regoper[]")]
  RegoperArray,
  #[display(fmt = "regoperator[]")]
  RegoperatorArray,
  #[display(fmt = "regclass[]")]
  RegclassArray,
  #[display(fmt = "regtype[]")]
  RegtypeArray,
  #[display(fmt = "record")]
  Record,
  #[display(fmt = "cstring")]
  Cstring,
  #[display(fmt = "any")]
  Any,
  #[display(fmt = "anyarray")]
  Anyarray,
  #[display(fmt = "void")]
  Void,
  #[display(fmt = "trigger")]
  Trigger,
  #[display(fmt = "languagehandler")]
  LanguageHandler,
  #[display(fmt = "internal")]
  Internal,
  #[display(fmt = "anyelement")]
  Anyelement,
  #[display(fmt = "record[]")]
  RecordArray,
  #[display(fmt = "anynonarray")]
  Anynonarray,
  #[display(fmt = "txidsnapshot[]")]
  TxidSnapshotArray,
  #[display(fmt = "uuid")]
  Uuid,
  #[display(fmt = "uuid[]")]
  UuidArray,
  #[display(fmt = "txidsnapshot")]
  TxidSnapshot,
  #[display(fmt = "fdwhandler")]
  FdwHandler,
  #[display(fmt = "pglsn")]
  PgLsn,
  #[display(fmt = "pglsn[]")]
  PgLsnArray,
  #[display(fmt = "tsmhandler")]
  TsmHandler,
  #[display(fmt = "pgndistinct")]
  PgNdistinct,
  #[display(fmt = "pgdependencies")]
  PgDependencies,
  #[display(fmt = "anyenum")]
  Anyenum,
  #[display(fmt = "tsvector")]
  TsVector,
  #[display(fmt = "tsquery")]
  Tsquery,
  #[display(fmt = "gtsvector")]
  GtsVector,
  #[display(fmt = "tsvector[]")]
  TsVectorArray,
  #[display(fmt = "gtsvector[]")]
  GtsVectorArray,
  #[display(fmt = "tsquery[]")]
  TsqueryArray,
  #[display(fmt = "regconfig")]
  Regconfig,
  #[display(fmt = "regconfig[]")]
  RegconfigArray,
  #[display(fmt = "regdictionary")]
  Regdictionary,
  #[display(fmt = "regdictionary[]")]
  RegdictionaryArray,
  #[display(fmt = "jsonb")]
  Jsonb,
  #[display(fmt = "jsonb[]")]
  JsonbArray,
  #[display(fmt = "anyrange")]
  AnyRange,
  #[display(fmt = "eventtrigger")]
  EventTrigger,
  #[display(fmt = "int4range")]
  Int4Range,
  #[display(fmt = "int4range[]")]
  Int4RangeArray,
  #[display(fmt = "numrange")]
  NumRange,
  #[display(fmt = "numrange[]")]
  NumRangeArray,
  #[display(fmt = "tsrange")]
  TsRange,
  #[display(fmt = "tsrange[]")]
  TsRangeArray,
  #[display(fmt = "tstzrange")]
  TstzRange,
  #[display(fmt = "tstzrange[]")]
  TstzRangeArray,
  #[display(fmt = "daterange")]
  DateRange,
  #[display(fmt = "daterange[]")]
  DateRangeArray,
  #[display(fmt = "int8range")]
  Int8Range,
  #[display(fmt = "int8range[]")]
  Int8RangeArray,
  #[display(fmt = "jsonpath")]
  Jsonpath,
  #[display(fmt = "jsonpath[]")]
  JsonpathArray,
  #[display(fmt = "regnamespace")]
  Regnamespace,
  #[display(fmt = "regnamespace[]")]
  RegnamespaceArray,
  #[display(fmt = "regrole")]
  Regrole,
  #[display(fmt = "regrole[]")]
  RegroleArray,
  #[display(fmt = "regcollation")]
  Regcollation,
  #[display(fmt = "regcollation[]")]
  RegcollationArray,
  #[display(fmt = "pgmcvlist")]
  PgMcvList,
  #[display(fmt = "pgsnapshot")]
  PgSnapshot,
  #[display(fmt = "pgsnapshot[]")]
  PgSnapshotArray,
  #[display(fmt = "xid8")]
  Xid8,
  #[display(fmt = "anycompatible")]
  Anycompatible,
  #[display(fmt = "anycompatiblearray")]
  Anycompatiblearray,
  #[display(fmt = "anycompatiblenonarray")]
  Anycompatiblenonarray,
  #[display(fmt = "anycompatiblerange")]
  AnycompatibleRange,
  #[serde(rename_all = "camelCase")]
  #[display(fmt = "{}", "_0")]
  Other(String),
}

#[cfg(test)]
mod tests {
  use super::*;
  use similar_asserts::assert_eq;

  #[test]
  fn serialize1() {
    let value = Type::Bit;

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(string, r#""bit""#);
      }
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn serialize2() {
    let value = Type::Int8;

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(string, r#""int8""#);
      }
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn serialize_other() {
    let value = Type::Other("wow".to_owned());

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(
          string,
          r#"{
  "other": "wow"
}"#
        );
      }
      Err(e) => panic!("{}", e),
    }
  }
}
