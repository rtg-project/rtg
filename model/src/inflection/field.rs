use inflector::Inflector;

pub fn inflect_name_from_sql_column_name(name: &str) -> String {
  Inflector::to_camel_case(name)
}

pub fn inflect_sql_column_name_from_name(name: &str) -> String {
  Inflector::to_snake_case(name)
}

pub fn inflect_graphql_field_name_from_name(name: &str) -> String {
  Inflector::to_camel_case(name)
}

pub fn inflect_graphql_order_by_asc_from_name(name: &str) -> String {
  format!("{}Asc", Inflector::to_camel_case(name))
}

pub fn inflect_graphql_order_by_desc_from_name(name: &str) -> String {
  format!("{}Desc", Inflector::to_camel_case(name))
}
