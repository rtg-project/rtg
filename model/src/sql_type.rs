// See rust-postgres
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

// See https://github.com/sfackler/rust-postgres/blob/fc10985f9fdf0903893109bc951fb5891539bf97/postgres-types/src/type_gen.rs
#[derive(Serialize, Deserialize, Debug, PartialEq, JsonSchema)]
#[serde(tag = "sqlType", rename_all = "camelCase")]
pub enum Type {
  Bool,
  Bytea,
  Char,
  Name,
  Int8,
  Int2,
  Int2Vector,
  Int4,
  Regproc,
  Text,
  Oid,
  Tid,
  Xid,
  Cid,
  OidVector,
  PgDdlCommand,
  Json,
  Xml,
  XmlArray,
  PgNodeTree,
  JsonArray,
  TableAmHandler,
  Xid8Array,
  IndexAmHandler,
  Point,
  Lseg,
  Path,
  Box,
  Polygon,
  Line,
  LineArray,
  Cidr,
  CidrArray,
  Float4,
  Float8,
  Unknown,
  Circle,
  CircleArray,
  Macaddr8,
  Macaddr8Array,
  Money,
  MoneyArray,
  Macaddr,
  Inet,
  BoolArray,
  ByteaArray,
  CharArray,
  NameArray,
  Int2Array,
  Int2VectorArray,
  Int4Array,
  RegprocArray,
  TextArray,
  TidArray,
  XidArray,
  CidArray,
  OidVectorArray,
  BpcharArray,
  VarcharArray,
  Int8Array,
  PointArray,
  LsegArray,
  PathArray,
  BoxArray,
  Float4Array,
  Float8Array,
  PolygonArray,
  OidArray,
  Aclitem,
  AclitemArray,
  MacaddrArray,
  InetArray,
  Bpchar,
  Varchar,
  Date,
  Time,
  Timestamp,
  TimestampArray,
  DateArray,
  TimeArray,
  Timestamptz,
  TimestamptzArray,
  Interval,
  IntervalArray,
  NumericArray,
  CstringArray,
  Timetz,
  TimetzArray,
  Bit,
  BitArray,
  Varbit,
  VarbitArray,
  Numeric,
  Refcursor,
  RefcursorArray,
  Regprocedure,
  Regoper,
  Regoperator,
  Regclass,
  Regtype,
  RegprocedureArray,
  RegoperArray,
  RegoperatorArray,
  RegclassArray,
  RegtypeArray,
  Record,
  Cstring,
  Any,
  Anyarray,
  Void,
  Trigger,
  LanguageHandler,
  Internal,
  Anyelement,
  RecordArray,
  Anynonarray,
  TxidSnapshotArray,
  Uuid,
  UuidArray,
  TxidSnapshot,
  FdwHandler,
  PgLsn,
  PgLsnArray,
  TsmHandler,
  PgNdistinct,
  PgDependencies,
  Anyenum,
  TsVector,
  Tsquery,
  GtsVector,
  TsVectorArray,
  GtsVectorArray,
  TsqueryArray,
  Regconfig,
  RegconfigArray,
  Regdictionary,
  RegdictionaryArray,
  Jsonb,
  JsonbArray,
  AnyRange,
  EventTrigger,
  Int4Range,
  Int4RangeArray,
  NumRange,
  NumRangeArray,
  TsRange,
  TsRangeArray,
  TstzRange,
  TstzRangeArray,
  DateRange,
  DateRangeArray,
  Int8Range,
  Int8RangeArray,
  Jsonpath,
  JsonpathArray,
  Regnamespace,
  RegnamespaceArray,
  Regrole,
  RegroleArray,
  Regcollation,
  RegcollationArray,
  PgMcvList,
  PgSnapshot,
  PgSnapshotArray,
  Xid8,
  Anycompatible,
  Anycompatiblearray,
  Anycompatiblenonarray,
  AnycompatibleRange,
  #[serde(rename_all = "camelCase")]
  Other {
    sql_type_name: String,
  },
}
