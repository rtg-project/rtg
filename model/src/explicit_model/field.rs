use super::sql_type;
use partial_struct::PartialStruct;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

#[derive(PartialStruct, Serialize, Deserialize, Debug, JsonSchema)]
#[serde(tag = "type", rename_all = "camelCase")]
#[partial(name = "ImplicitField")]
#[partial_attribute(derive(Serialize, Deserialize, Debug, JsonSchema))]
#[partial_attribute(serde(tag = "type", rename_all = "camelCase"))]
pub enum ExplicitField {
  #[serde(rename_all = "camelCase")]
  ScalarDatabaseColumn {
    name: String,
    nullable: bool,
    // #[serde(flatten)]
    sql_type: sql_type::Type,
    sql_column_name: String,
    graphql_field_name: String,
    graphql_type_name: String,
    graphql_order_by_asc: String,
    graphql_order_by_desc: String,
  },
}

// Tests
#[cfg(test)]
mod tests {
  use super::*;
  use similar_asserts::assert_eq;

  #[test]
  fn serialize() {
    let value = ExplicitField::ScalarDatabaseColumn {
      name: "drone".to_owned(),
      nullable: false,
      sql_type: sql_type::Type::Text,
      sql_column_name: "drone_col".to_owned(),
      graphql_field_name: "drone".to_owned(),
      graphql_type_name: "String".to_owned(),
      graphql_order_by_asc: "drone_ASC".to_owned(),
      graphql_order_by_desc: "drone_DESC".to_owned(),
    };

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(
          string,
          r#"{
  "type": "scalarDatabaseColumn",
  "name": "drone",
  "nullable": false,
  "sqlType": "text",
  "sqlColumnName": "drone_col",
  "graphqlFieldName": "drone",
  "graphqlTypeName": "String",
  "graphqlOrderByAsc": "drone_ASC",
  "graphqlOrderByDesc": "drone_DESC"
}"#
        );
      }
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn serialize_other() {
    let value = ExplicitField::ScalarDatabaseColumn {
      name: "drone".to_owned(),
      nullable: false,
      sql_type: sql_type::Type::Other("Yoo".to_owned()),
      sql_column_name: "drone_col".to_owned(),
      graphql_field_name: "drone".to_owned(),
      graphql_type_name: "String".to_owned(),
      graphql_order_by_asc: "drone_ASC".to_owned(),
      graphql_order_by_desc: "drone_DESC".to_owned(),
    };

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(
          string,
          r#"{
  "type": "scalarDatabaseColumn",
  "name": "drone",
  "nullable": false,
  "sqlType": {
    "other": "Yoo"
  },
  "sqlColumnName": "drone_col",
  "graphqlFieldName": "drone",
  "graphqlTypeName": "String",
  "graphqlOrderByAsc": "drone_ASC",
  "graphqlOrderByDesc": "drone_DESC"
}"#
        );
      }
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn deserialize() {
    let data = r#"
      {
        "name": "id",
        "nullable": false,
        "type": "scalarDatabaseColumn",
        "sqlType": "text",
        "sqlColumnName": "id",
        "graphqlFieldName": "id",
        "graphqlTypeName": "String",
        "graphqlOrderByAsc": "idAsc",
        "graphqlOrderByDesc": "idDesc"
      }
      "#;

    match serde_json::from_str(data) {
      Ok(field) => match field {
        ExplicitField::ScalarDatabaseColumn {
          name,
          nullable,
          sql_type,
          sql_column_name,
          graphql_field_name: _,
          graphql_type_name: _,
          graphql_order_by_asc: _,
          graphql_order_by_desc: _,
        } => {
          assert_eq!(name, "id");
          assert_eq!(nullable, false);
          assert_eq!(sql_type, sql_type::Type::Text);
          assert_eq!(sql_column_name, "id");
        }
      },
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn deserialize_other() {
    let data = r#"
      {
        "name": "id",
        "type": "scalarDatabaseColumn",
        "nullable": true,
        "sqlType": {
          "other": "Hii"
        },
        "sqlColumnName": "id",
        "graphqlFieldName": "id",
        "graphqlTypeName": "String",
        "graphqlOrderByAsc": "idAsc",
        "graphqlOrderByDesc": "idDesc"
      }
      "#;

    match serde_json::from_str(data) {
      Ok(field) => match field {
        ExplicitField::ScalarDatabaseColumn {
          name,
          nullable,
          sql_type,
          sql_column_name,
          graphql_field_name: _,
          graphql_type_name: _,
          graphql_order_by_asc: _,
          graphql_order_by_desc: _,
        } => {
          assert_eq!(name, "id");
          assert_eq!(nullable, true);
          assert_eq!(sql_type, sql_type::Type::Other("Hii".to_owned()));
          assert_eq!(sql_column_name, "id");
        }
      },
      Err(e) => panic!("{}", e),
    }
  }
}
