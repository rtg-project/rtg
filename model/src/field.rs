use super::sql_type;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum Field {
  #[serde(rename_all = "camelCase")]
  ScalarDatabaseColumn {
    name: String,
    #[serde(flatten)]
    sql_type: sql_type::Type,
    sql_column_name: String,
    graphql_field_name: String,
    graphql_type_name: String,
    graphql_order_by_asc: String,
    graphql_order_by_desc: String,
  },
}

// Tests
#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn serialize() {
    let value = Field::ScalarDatabaseColumn {
      name: "drone".to_string(),
      sql_type: sql_type::Type::Text,
      sql_column_name: "drone_col".to_string(),
      graphql_field_name: "drone".to_string(),
      graphql_type_name: "String".to_string(),
      graphql_order_by_asc: "drone_ASC".to_string(),
      graphql_order_by_desc: "drone_DESC".to_string(),
    };

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(
          string,
          r#"{
  "type": "scalarDatabaseColumn",
  "name": "drone",
  "sqlType": "text",
  "sqlColumnName": "drone_col",
  "graphqlFieldName": "drone",
  "graphqlTypeName": "String",
  "graphqlOrderByAsc": "drone_ASC",
  "graphqlOrderByDesc": "drone_DESC"
}"#
        );
      }
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn serialize_other() {
    let value = Field::ScalarDatabaseColumn {
      name: "drone".to_string(),
      sql_type: sql_type::Type::Other {
        sql_type_name: "Yoo".to_string(),
      },
      sql_column_name: "drone_col".to_string(),
      graphql_field_name: "drone".to_string(),
      graphql_type_name: "String".to_string(),
      graphql_order_by_asc: "drone_ASC".to_string(),
      graphql_order_by_desc: "drone_DESC".to_string(),
    };

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(
          string,
          r#"{
  "type": "scalarDatabaseColumn",
  "name": "drone",
  "sqlType": "other",
  "sqlTypeName": "Yoo",
  "sqlColumnName": "drone_col",
  "graphqlFieldName": "drone",
  "graphqlTypeName": "String",
  "graphqlOrderByAsc": "drone_ASC",
  "graphqlOrderByDesc": "drone_DESC"
}"#
        );
      }
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn deserialize() {
    let data = r#"
      {
        "name": "id",
        "type": "scalarDatabaseColumn",
        "sqlType": "text",
        "sqlColumnName": "id",
        "graphqlFieldName": "id",
        "graphqlTypeName": "String",
        "graphqlOrderByAsc": "idAsc",
        "graphqlOrderByDesc": "idDesc"
      }
      "#;

    match serde_json::from_str(data) {
      Ok(field) => match field {
        Field::ScalarDatabaseColumn {
          name,
          sql_type,
          sql_column_name,
          graphql_field_name: _,
          graphql_type_name: _,
          graphql_order_by_asc: _,
          graphql_order_by_desc: _,
        } => {
          assert_eq!(name, "id");
          assert_eq!(sql_type, sql_type::Type::Text);
          assert_eq!(sql_column_name, "id");
        }
      },
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn deserialize_other() {
    let data = r#"
      {
        "name": "id",
        "type": "scalarDatabaseColumn",
        "sqlType": "other",
        "sqlTypeName": "Hii",
        "sqlColumnName": "id",
        "graphqlFieldName": "id",
        "graphqlTypeName": "String",
        "graphqlOrderByAsc": "idAsc",
        "graphqlOrderByDesc": "idDesc"
      }
      "#;

    match serde_json::from_str(data) {
      Ok(field) => match field {
        Field::ScalarDatabaseColumn {
          name,
          sql_type,
          sql_column_name,
          graphql_field_name: _,
          graphql_type_name: _,
          graphql_order_by_asc: _,
          graphql_order_by_desc: _,
        } => {
          assert_eq!(name, "id");
          assert_eq!(
            sql_type,
            sql_type::Type::Other {
              sql_type_name: "Hii".to_string(),
            }
          );
          assert_eq!(sql_column_name, "id");
        }
      },
      Err(e) => panic!("{}", e),
    }
  }
}
