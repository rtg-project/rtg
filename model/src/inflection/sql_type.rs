use super::inflection_error;
use crate::explicit_model::Type;

pub fn inflect_sql_type_from_graphql_type(
  graphql_type: &str,
  is_array: bool,
) -> Result<Type, inflection_error::InflectionError> {
  match is_array {
    false => match graphql_type {
      "String" => Ok(Type::Text),
      "Boolean" => Ok(Type::Bool),
      "SmallInt" => Ok(Type::Int2),
      "Int" => Ok(Type::Int4),
      "BigInt" => Ok(Type::Int8),
      "Float" => Ok(Type::Float4),
      "Double" => Ok(Type::Float8),
      "Numeric" => Ok(Type::Numeric),
      "DateTime" => Ok(Type::Timestamptz),
      "Json" => Ok(Type::Jsonb),
      "Bytes" => Ok(Type::Bytea),
      type_name => Err(inflection_error::InflectionError::UnsupportedGraphQLScalar(
        type_name.to_string(),
      )),
    },
    true => match graphql_type {
      "String" => Ok(Type::TextArray),
      "Boolean" => Ok(Type::BoolArray),
      "SmallInt" => Ok(Type::Int2Array),
      "Int" => Ok(Type::Int4Array),
      "BigInt" => Ok(Type::Int8Array),
      "Float" => Ok(Type::Float4Array),
      "Double" => Ok(Type::Float8Array),
      "Numeric" => Ok(Type::NumericArray),
      "DateTime" => Ok(Type::TimestamptzArray),
      "Json" => Ok(Type::JsonbArray),
      "Bytes" => Ok(Type::ByteaArray),
      type_name => Err(inflection_error::InflectionError::UnsupportedGraphQLScalar(
        type_name.to_string(),
      )),
    },
  }
}

pub fn inflect_graphql_type_from_sql_type(
  sql_type: &Type,
) -> Result<(String, bool), inflection_error::InflectionError> {
  match sql_type {
    Type::Text => Ok(("String".to_string(), false)),
    Type::Bool => Ok(("Boolean".to_string(), false)),
    Type::Int2 => Ok(("SmallInt".to_string(), false)),
    Type::Int4 => Ok(("Int".to_string(), false)),
    Type::Int8 => Ok(("BigInt".to_string(), false)),
    Type::Float4 => Ok(("Float".to_string(), false)),
    Type::Float8 => Ok(("Double".to_string(), false)),
    Type::Numeric => Ok(("Numeric".to_string(), false)),
    Type::Timestamptz => Ok(("DateTime".to_string(), false)),
    Type::Jsonb => Ok(("Json".to_string(), false)),
    Type::Json => Ok(("Json".to_string(), false)),
    Type::Bytea => Ok(("Bytes".to_string(), false)),
    Type::TextArray => Ok(("String".to_string(), true)),
    Type::BoolArray => Ok(("Boolean".to_string(), true)),
    Type::Int2Array => Ok(("SmallInt".to_string(), true)),
    Type::Int4Array => Ok(("Int".to_string(), true)),
    Type::Int8Array => Ok(("BigInt".to_string(), true)),
    Type::Float4Array => Ok(("Float".to_string(), true)),
    Type::Float8Array => Ok(("Double".to_string(), true)),
    Type::NumericArray => Ok(("Numeric".to_string(), true)),
    Type::TimestamptzArray => Ok(("DateTime".to_string(), true)),
    Type::JsonbArray => Ok(("Json".to_string(), true)),
    Type::JsonArray => Ok(("Json".to_string(), true)),
    Type::ByteaArray => Ok(("Bytes".to_string(), true)),
    type_name => Err(inflection_error::InflectionError::UnsupportedSQLType(
      format!("{:?}", type_name),
    )),
  }
}
