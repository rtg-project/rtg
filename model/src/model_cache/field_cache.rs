use crate::explicit_model::field::ExplicitField;
use serde::{Deserialize, Serialize};
use std::rc::Rc;

#[derive(Serialize, Deserialize, Debug)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum FieldCache {
  #[serde(rename_all = "camelCase")]
  ScalarDatabaseColumn { field: Rc<ExplicitField> },
}

impl FieldCache {
  // Another associated function, taking two arguments:
  pub fn new(field: Rc<ExplicitField>) -> FieldCache {
    match *field {
      ExplicitField::ScalarDatabaseColumn { .. } => {
        return FieldCache::ScalarDatabaseColumn {
          field: Rc::clone(&field),
        };
      }
    };
  }
}

// Tests
#[cfg(test)]
mod tests {
  use super::*;
  use crate::explicit_model::sql_type;
  use insta::{assert_debug_snapshot, assert_json_snapshot};

  #[test]
  fn constructor() {
    let value = Rc::new(ExplicitField::ScalarDatabaseColumn {
      name: "id".to_string(),
      sql_type: sql_type::Type::Text,
      sql_column_name: "id_col".to_string(),
      graphql_field_name: "id".to_string(),
      graphql_type_name: "String".to_string(),
      graphql_order_by_asc: "id_ASC".to_string(),
      graphql_order_by_desc: "id_DESC".to_string(),
    });

    assert_debug_snapshot!(FieldCache::new(Rc::clone(&value)));
    assert_json_snapshot!(
      serde_json::to_string_pretty(&FieldCache::new(Rc::clone(&value))).unwrap()
    );
  }
}
