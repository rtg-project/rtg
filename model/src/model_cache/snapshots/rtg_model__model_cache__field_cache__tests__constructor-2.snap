---
source: model/src/model_cache/field_cache.rs
expression: "serde_json::to_string_pretty(&FieldCache::new(Rc::clone(&value))).unwrap()"

---
"{\n  \"type\": \"scalarDatabaseColumn\",\n  \"field\": {\n    \"type\": \"scalarDatabaseColumn\",\n    \"name\": \"id\",\n    \"nullable\": false,\n    \"sqlType\": \"text\",\n    \"sqlColumnName\": \"id_col\",\n    \"graphqlFieldName\": \"id\",\n    \"graphqlTypeName\": \"String\",\n    \"graphqlOrderByAsc\": \"id_ASC\",\n    \"graphqlOrderByDesc\": \"id_DESC\"\n  }\n}"
